src/codelabs/null-safety.md

```diff
@@ -0,0 +1,362 @@
+```dart:run-dartpad:ga_id-nonnullable_type:null_safety-true
+void main() {
+  String text;
+
+  //if (DateTime.now().hour < 12) {
+  //  text = "It's morning! Let's make aloo paratha!";
+  //} else {
+  //  text = "It's afternoon! Let's make biryani!";
+  //}
+
+  print(text);
+  print(text.length);
+}
+```
+
+## New null-aware operators: ? and !
+
+Null safety introduces two operators, ? and !. Here's how you use them:
+
+-   If you're creating a variable that can have the value null, add a
+    question mark (?) to the variable's type declaration. Types are
+    non-nullable by default and made nullable by adding ?.
+-   If you're accessing a property and the operand might be null, use the
+    null-aware version member access operator (?.).
+-   If you're sure that an expression with a nullable type isn't null, you can 
+    add a postfix exclamation mark (!) to make Dart treat it as non-nullable.
+
+### Example: Conditional access
+
+Conditional access is a handy way to tighten up code that needs to read
+properties that could be null. For example, the following expression evaluates
+to the value of `b` as long as `a` isn't null. If `a` is null, then the
+expression evaluates to null:
+
+  a?.b
+
+Without null safety, *two* conditional access operators are needed for the
+following code to work: one after `big` and one after `little`. With null
+safety, conditional access can short-circuit, so this expression only requires a
+single `?`:
+
+<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/conditional_access.dart" replace="/big\?.little/big.little/g"?>
+```dart:run-dartpad:ga_id-nonnullable_type:null_safety-true
+class BigThing {
+  LittleThing little = LittleThing();
+}
+
+class LittleThing {
+  int fetchInt() => 12;
+}
+
+void main() {
+  BigThing? big = BigThing();
+
+  print('The value is:');
+  print(big.little.fetchInt());
+}
+```
+
+### Example: The assertion operator `!`
+
+If you'd like to assign a nullable expression to a variable that's
+non-nullable, you can use the assertion operator (the exclamation point: `!`).
+By adding `!` just after the expression, you tell Dart that the value won't be
+null, and that it's safe to assign it to a non-nullable variable. If you're
+wrong, Dart throws an exception.
+
+In the example below, try adding exclamation points to correct the
+three broken assignments:
+
+<?code-excerpt "../null_safety_examples/null_safety_codelab/bin/assertion_operator.dart" replace="/first!/first/g; /!.abs/.abs/g"?>
+```dart:run-dartpad:ga_id-nonnullable_type:null_safety-true
+int? couldReturnNullButDoesnt() => -3;
+
+void main() {
+  int? couldBeNullButIsnt = 1;
+  List<int?> listThatCouldHoldNulls = [2, null, 4];
+
+  int a = couldBeNullButIsnt;
+  int b = listThatCouldHoldNulls.first; // first item in the list
+  int c = couldReturnNullButDoesnt().abs(); // absolute value
+
+  print('a is $a.');
+  print('b is $b.');
+  print('c is $c.');
+}
+```
+
```

Comment:
This seems like a new section on type promotion. Should we add a section header move the first paragraph from the "Type promotion" exercise up here?

Reply:
done. LMK if that's what you had in mind.